File: TypeScript/ex3.ts
interface Vitesse {
    Vmax():string;
};

abstract class Vehicule{
    private _name:string;
    protected vitesse:string;

    protected set name(v : string) {
        this._name = v;
    }
    
    protected get name() : string {
        return this._name;
    }

    constructor(name : string){
        this.name = name;
    }

}

class voiture extends Vehicule implements Vitesse
{  
    constructor(name : string){
        super(name);
        this.vitesse = " 130 km/h";
    }

    public Vmax() :string{
        return this.name + this.vitesse;
    }
    
};

class moto extends Vehicule implements Vitesse
{  
    constructor(name : string){
        super(name);
        this.vitesse = " 200 km/h";
    }

    public Vmax() :string{
        return this.name + this.vitesse;
    }
    
};

class camion extends Vehicule implements Vitesse
{  
    constructor(name : string){
        super(name);
        this.vitesse = " 80 km/h";
    }

    public Vmax() :string{
        return this.name + this.vitesse;
    }
    
};

function ex3() : void
{
    let tab : Array<Vitesse> = new Array<Vitesse>();
    tab.push(new voiture("C3"));
    tab.push(new moto("R1"));
    tab.push(new camion("Man"));
    
    for (let i = 0; i < 3; i++)
    {
        console.log(tab[i].Vmax());
    }
    
    return;
}
ex3();

File: rust/src/ex3.rs
// src : ChatGPT

trait Vitesse {
    fn vmax(&self) -> String;
}

struct Voiture {
    name: String,
    vitesse: String,
}

impl Vitesse for Voiture {
    fn vmax(&self) -> String {
        format!("{}{}", self.name, self.vitesse)
    }
}

struct Moto {
    name: String,
    vitesse: String,
}

impl Vitesse for Moto {
    fn vmax(&self) -> String {
        format!("{}{}", self.name, self.vitesse)
    }
}

struct Camion {
    name: String,
    vitesse: String,
}

impl Vitesse for Camion {
    fn vmax(&self) -> String {
        format!("{}{}", self.name, self.vitesse)
    }
}

pub fn ex3(){
    let taille_tab = 3;
    let mut tab: Vec<Box<dyn Vitesse>> = Vec::new();
    
    tab.push(Box::new(Voiture { name: String::from("C3"), vitesse: String::from(" 130 km/h") }));
    tab.push(Box::new(Moto { name: String::from("R1"), vitesse: String::from(" 200 km/h") }));
    tab.push(Box::new(Camion { name: String::from("Man"), vitesse: String::from(" 80 km/h") }));

    for i in 0..taille_tab {
        println!("{}", tab[i].vmax());
    }
}